fold/fold.go

```diff
@@ -0,0 +1,895 @@
+package fold
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/TimothyStiles/poly/transform"
+)
+
+// Fold the DNA sequence and return the lowest free energy score.
+//
+// Based on the approach described in:
+// Zuker and Stiegler, 1981
+// https://www.ncbi.nlm.nih.gov/pmc/articles/PMC326673/pdf/nar00394-0137.pdf
+//
+// If the sequence is 50 or more bp long, "isolated" matching bp
+// are ignored in V(i,j). This is based on an approach described in:
+// Mathews, Sabina, Zuker and Turner, 1999
+// https://www.ncbi.nlm.nih.gov/pubmed/10329189
+// Args:
+//
+//	seq: The sequence to Fold
+//	temp: The temperature the Fold takes place in, in Celsius
+//
+// Returns a slice of NucleicAcidStructure with the energy and description,
+// i.e. stacks, bulges, hairpins, etc.
+func Fold(seq string, temp float64) ([]NucleicAcidStructure, error) {
+	fc, err := NewFoldingContext(seq, temp)
+	if err != nil {
+		return nil, fmt.Errorf("error creating folding context: %w", err)
+	}
+
+	// get the minimum free energy structure out of the cache
+	return Traceback(0, len(seq)-1, fc), nil
+}
+
+// Fold the sequence and return just the delta G of the structure
+// Args:
+//
+//	seq: The sequence to fold
+//	temp: The temperature to fold at
+//
+// Returns: the minimum free energy of the folded sequence
+func FoldEnergy(seq string, temp float64) (float64, error) {
+	structs, err := Fold(seq, temp)
+	if err != nil {
+		return 0, fmt.Errorf("error folding: %w", err)
+	}
+	sumE := 0.0
+	for _, s := range structs {
+		sumE += s.E
+	}
+	return RoundFloat(sumE, 2), nil
+}
+
+// Get the dot bracket notation for a secondary structure.
+// Args:
+//
+//	structs: A list of NucleicAcidStructure, usually from the fold function
+//
+// # Returns the dot-bracket notation of the secondary structure
+//
+// Dot-bracket notation, consisting in a balanced parentheses string composed
+// by a three-character alphabet {.,(,)}, that can be unambiguously converted
+// in the RNA secondary structure. See example_test.go for a small example.
+func DotBracket(structs []NucleicAcidStructure) string {
+	maxj := 0
+	for _, s := range structs {
+		for _, ij := range s.Inner {
+			if ij.End > maxj {
+				maxj = ij.End
+			}
+		}
+	}
+	maxj += 1
+	result := make([]byte, maxj)
+	for i := range result {
+		result[i] = '.'
+	}
+	for _, s := range structs {
+		if len(s.Inner) == 1 {
+			ij := s.Inner[0]
+			result[ij.Start] = '('
+			result[ij.End] = ')'
+		}
+	}
+	return string(result)
+}
+
+// Find and return the lowest free energy structure in the subsequence starting
+// at i and terminating at j.
+//
+// Figure 2B in Zuker and Stiegler, 1981
+// Args:
+//
+//		seq: The sequence being folded
+//		i: The start index
+//		j: The end index (inclusive)
+//	 fc: The FoldingContext for this sequence
+//
+// Returns the free energy for the subsequence from i to j
+func W(i, j int, fc FoldContext) (NucleicAcidStructure, error) {
```

Comment:
should this be unexported?

Reply:
Depends on the use for this package and its usefulness outside the folding. 

We could start not exporting anything that is not needed in a `cmd` and then export in case it's needed. This way we keep the API surface small and less prone to breaking changes?